joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -269,19 +349,26 @@ def loop(self):
                     joint = self.free_joints[parent]
 
                 if has_position and 'position' in joint:
-                    msg.position[i] = joint['position'] * factor + offset
+                    msg.position[i] = float(joint['position']) * factor + offset
                 if has_velocity and 'velocity' in joint:
-                    msg.velocity[i] = joint['velocity'] * factor
+                    msg.velocity[i] = float(joint['velocity']) * factor
                 if has_effort and 'effort' in joint:
-                    msg.effort[i] = joint['effort']
+                    msg.effort[i] = float(joint['effort'])
 
             if msg.name or msg.position or msg.velocity or msg.effort:
                 # Only publish non-empty messages
                 self.pub.publish(msg)
-            try:
-                r.sleep()
-            except rospy.exceptions.ROSTimeMovedBackwardsException:
-                pass
+
+            # Spin to make sure we service any subscriptions
+            rclpy.spin_once(self.node, timeout_sec=0)
```

Comment:
This only executes a single callback. If multiple things are ready, then acting on that data will be delayed by the sleep below. I strongly recommend using a timer instead of looping.

Reply:
Fixed in 2917803.